### ISSUE TITLE
Fails to Install on M1 Mac because of -march=native flag

### ISSUE THREAD

**Issue by shamanskyh** [irmen/pyminiaudio#30]
Hi, when I try to install pyminiaudio on an M1 Mac mini by running `pip install miniaudio`, I get the following error, which appears to be caused by the `-march=native` flag:
```
ERROR: Command errored out with exit status 1:
     command: /Library/Developer/CommandLineTools/usr/bin/python3 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5/setup.py'"'"'; __file__='"'"'/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-record-ox8ldrmt/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /Users/harry/Library/Python/3.8/include/python3.8/miniaudio
         cwd: /private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5/
    Complete output (15 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.14.6-arm64-3.8
    copying miniaudio.py -> build/lib.macosx-10.14.6-arm64-3.8
    running build_ext
    generating cffi module 'build/temp.macosx-10.14.6-arm64-3.8/_miniaudio.c'
    creating build/temp.macosx-10.14.6-arm64-3.8
    building '_miniaudio' extension
    creating build/temp.macosx-10.14.6-arm64-3.8/build
    creating build/temp.macosx-10.14.6-arm64-3.8/build/temp.macosx-10.14.6-arm64-3.8
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -DMA_NO_GENERATION=1 -I/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5 -I/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5 -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/include/python3.8 -c build/temp.macosx-10.14.6-arm64-3.8/_miniaudio.c -o build/temp.macosx-10.14.6-arm64-3.8/build/temp.macosx-10.14.6-arm64-3.8/_miniaudio.o -g1 -O3 -ffast-math -mtune=native -march=native
    clang: error: the clang compiler does not support '-march=native'
    error: command 'clang' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Library/Developer/CommandLineTools/usr/bin/python3 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5/setup.py'"'"'; __file__='"'"'/private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-install-xr993ftf/miniaudio_41126c03fa9b4343962ac371bf368fd5/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/0z/cx87zyvn7ssfpp4fc7bff4rw0000gn/T/pip-record-ox8ldrmt/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /Users/harry/Library/Python/3.8/include/python3.8/miniaudio Check the logs for full command output.
```

**Comment by fidesachates** [comment]
I'm having the same issue on my intel macbook pro.

2.6 GHz 6-Core Intel Core i7
MacBook Pro (16-inch, 2019)
Big Sur 11.5.2

**Comment by postlund** [comment]
From what I have gathered, `-march=native` (possibly also `-mtune=native`) is not supported on M1 so they should be omitted. Maybe also for other platforms as well?

@irmen I usually don't like to ping people personally, but I see need to fix this without creating another fork. If someone came up with a PR, can you review and possibly make a new release to fix this problem? I expect #32 to have the same root cause.

**Comment by irmen** [comment]
What would the equivalent compiler flag be for M1 platform?

**Comment by postlund** [comment]
It seems llvm doesn't support M1 yet (or at least is not included in the current version), so "native" won't work as no definition exists for the platform. Also looks like cffi does not support ARM-on-Mac for the time being. I have not verified if that is still valid.

https://foss.heptapod.net/pypy/cffi/-/issues/464#note_147550

Anyway, the M1 seems to be defined in the same way as A14 right now:

https://github.com/llvm/llvm-project/blob/26b803177456112b4c6c2d30b7c3dc4d439af7e2/llvm/include/llvm/Support/AArch64TargetParser.def#L197

A potentially more viable solution for now might be to pick a more generic aarch64 platform and not lock down tightly to the M1. It will produce less effective code, but likely good enough. As the M1 uses the ARMv8.4-A instruction set, this is probably what we want to use:

https://github.com/llvm/llvm-project/blob/26b803177456112b4c6c2d30b7c3dc4d439af7e2/llvm/include/llvm/Support/AArch64TargetParser.def#L37

To detect platform, we can use this:

https://docs.python.org/3/library/platform.html#platform.machine

It will however only tell us if current system is aarch64 (or something else of course). If python can figure out CPU name for it, we can use this:

https://docs.python.org/3/library/platform.html#platform.processor

Otherwise I guess we need to try with `-march=native` set and fall back to the solutions above on failure.

One solution mentioned in the referenced issue in pyatv is to just remove the flags. I don't know what implications that would give in all different cases, but might be worth looking into.

**Comment by irmen** [comment]
Yup, removing the flag is a very easy solution and maybe the resulting output is "fast enough" on most systems. But as far as I know, this would result on the X86-64 platform (likely the largest) that the compiler will no longer use any of the advanced cpu instructions. For some operations this could result in a large speed decrease.  But I have no benchmarks for any of this to back it all up, these remarks are based on observations made elsewhere

**Comment by postlund** [comment]
I think that defaulting to no optimizations just in case is not a good idea, we should be more specific here. A simple solution I guess could be something pseudo-like:

```python
import platform

libraries = []
compiler_args = []
if os.name == "posix": 
    libraries = ["m", "pthread", "dl"]
    compiler_args = ["-g1", "-O3", "-ffast-math"]

    if platform.system() == "Darwin" and platform.machine() == "aarch64":  # Apple Silicon
        compiler_args += ["-mtune=apple-latest", "-march=apple-latest"]
    else:
        compiler_args += ["-mtune=native", "-march=native"]
```

My clang reports `apple-latest` as valid, but we might we to experiment with that. Would have been a lot easier with an M1 at hand.

@shamanskyh Maybe you can play around with various values for `-mtune` and `-march` and see what happens?

**Comment by irmen** [comment]
I added a variant of the above code in 5772fb217b822591c1044c1bd5a0fdd60453b109  can you please check if the current master branch now builds correctly?

**Comment by MartySkinner** [comment]
> 5772fb2

@irmen The platform.machine on my M1 Mac system is "arm64", not "aarch64".

**Comment by irmen** [comment]
fixed that.

**Comment by MartySkinner** [comment]
Sadly, even though `clang --print-supported-cpus` shows "apple-latest" as one of its supported CPUs, it isn't understood as an argument for `-mtune`. And the same goes for `-march` — it doesn't like "apple-latest" (from the errors I was seeing while playing around with it, I couldn't find anything to even suggest `-march` exists at all except the man page.)

So without searching for working examples from other packages, I don't know what to suggest for the Mac M1 arguments except perhaps nothing at all for now?

> Apple clang version 12.0.5 (clang-1205.0.22.9)
> Target: arm64-apple-darwin20.6.0
> Thread model: posix
> InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Comment by irmen** [comment]
@MartySkinner  can you try if "-mcpu=apple-latest" works?  If that also fails, we'll revert to just leaving the option out altogether.

**Comment by MartySkinner** [comment]
```
% clang foo.c -o foo.o -mcpu=apple-latest
clang: error: the clang compiler does not support '-mcpu=apple-latest'
```

**Comment by fidesachates** [comment]
As I said above, I'm having this issue, but on I'm on an intel macbook pro. Do you want me to create a separate issue? 

I tried installing https://github.com/irmen/pyminiaudio/commit/5772fb217b822591c1044c1bd5a0fdd60453b109, but that fails with the same error (which was expected as the change still uses `-march=native`

**Comment by irmen** [comment]
@fidesachates I don't have access to a mac so can you experiment with the flags on your system to make it work, and let me know  what apple requires

**Comment by eugenet8k** [comment]
Hi guys, I got the same error on my Intel Mac Mini with macOS v11.5.2 while installing the latest homeassistant [v2021.9.7](https://github.com/home-assistant/core/releases/tag/2021.9.7).

While you are reviewing the issue and before the fix is published, would you please give a suggestion about possible ways to immediately solve it to unblock the installation of Homeassistant software? Is it possible to build it without mentioned params manually and make the build accessible somehow to homeassistant?

Thank you for any advice.

**Comment by eugenet8k** [comment]
@irmen thank you, I was able to build and install the wheel locally. I need to point out that in the current `build_ffi_module.py` code:

```
# determine appropriate compiler tuning options
if platform.system() == "Darwin" and platform.machine() == "aarch64":
    # Apple Silicon
    compiler_args += ["-mtune=apple-latest", "-march=apple-latest"]
elif "ppc64" in platform.system():
    compiler_args += ["-mcpu=native"]
else:
    compiler_args += ["-mtune=native", "-march=native"]
```

my case (Intel CPU) didn't match any of the conditions, because my `platform.machine()` is `x86_64` and not `aarch64`. I updated the last line into just `compiler_args += []` to make the build pass.

**Comment by irmen** [comment]
well, I just removed all compiler tune options for any apple. 
32b2eb71290199bc8fe687322e59b7c83b177421 should now fix this build problem.
Feel free to reopen if not

**Comment by eugenet8k** [comment]
Thank you @irmen, I was able to build on my Intel Mac with the latest master. Please issue a new release version so other software like (homeassistant) could update the dependecy.

**Comment by irmen** [comment]
yeah I may do so soon, There were some other issues open that I wanted to wait a bit more to hopefully get an answer on

**Comment by irmen** [comment]
@eugenet8k 1.45 was just released
